tests/ds_unselect_rules.sh

```diff
@@ -10,11 +10,11 @@ DS=$1
 UNSELECT_LIST=$2
 
 
-printf "Copying $DS file to /tmp\n"
+printf "Copying %s file to /tmp\n" "$DS"
 cp $DS /tmp || exit 1
 DS="/tmp/$(basename $DS)"
 
-printf "Unselecting rules listed in the $UNSELECT_LIST from the $DS\n"
+printf "Unselecting rules listed in the %s from the %s\n" "$UNSELECT_LIST" "$DS"
```

Comment:
ditto

Reply:
printf "Copying %s file to /tmp\n" "$DS" 
and
printf 'Copying %s file to /tmp\n' "$DS" 
The results of these two methods are the same.   
And I don't understand what "the first argument" means.
Look forward to your reply.

Reply:
Sure, see double quotes section of the Bash manual: https://www.gnu.org/software/bash/manual/html_node/Double-Quotes.html, especially the backslash-related part.

Reply:
“\n” is interpreted as special character by printf, so there is no need to use  a backslash "\".
 You can see   https://www.computerhope.com/unix/uprintf.htm.
In addition，"printf "Copying %s file to /tmp\n" "$DS" "  what I modified  is based on the usage of printf，although the results are the same.

Reply:
Please read the linked piece again. It reads:

> Enclosing characters in double quotes (‘"’) preserves the literal value of all characters within the quotes, with the exception of ‘$’, ‘`’, ‘\’ 
> ...
> The backslash retains its special meaning only when followed by one of the following characters: ‘$’, ‘`’, ‘"’, ‘\’, or newline.

Behavior of backslash followed by literal `n`, i.e. `\n`, is therefore not defined, and `shellcheck` actually complains about that. You are right about what `printf` does, but it has to go through the shell processing first. If you use single quotes, there is no shell processing, so `\n` gets straight to printf.

Reply:
OK，I will use single quotes.
In addition，
`Thebackslash retains its special meaning only when followed by one of the following characters: ‘$’, ‘`’, ‘"’, ‘\’, or newline. Within double quotes, backslashes that are followed by one of these characters are removed. `
The character "newline" means “\n”, it's a whole.  According to the second sentence , I don't think there is any need to add \ in front of  \n.

Reply:
`\n` is backslash followed by letter `n`, not a newline. By newline they mean literal newline. As newline is difficult to show in the manual, they describe it by that word.